File: 0301_0400/0397/0397.py
with open("input.txt", "r") as f, open("output.txt", "w") as q:
    s = f.readline().strip()
    n, a, b = len(s), min(s), max(s)
    if a != b:
        l = [(i, 0) for i in range(n) if s[i] == a]
        l.extend((i, 1) for i in range(n) if s[i] == b)
        l.sort()
        for i in range(1, len(l)):
            if l[i][1] != l[i - 1][1] and l[i][0] - l[i - 1][0] < n:
                n, d = l[i][0] - l[i - 1][0], i - 1
        a = s[l[d][0] : l[d + 1][0] + 1]
    q.write(a)

File: 0301_0400/0397/0397.c
#include <stdio.h>
#include <string.h>

int main()
{
	FILE *f = fopen("input.txt", "r"), *q = fopen("output.txt", "w");
	int n, r, l, max, min, b;
	char s[200001];

	fscanf(f, "%s", s);
	r = n = strlen(s);
	max = min = s[0];
	for (int i = 0; i < n; ++i)
		if (s[i] > max)
			max = s[i];
		else if (s[i] < min)
			min = s[i];
	if (min < max)
	{
		for (l = 0; s[l] != min && s[l] != max; ++l);
		for (int i = 0; i < n; ++i)
			if (s[i] == min || s[i] == max)
			{
				if (s[i] != s[l] && i - l < r)
					r = i - l, b = l;
				l = i;
			}
		fprintf(q, "%.*s", r + 1, s + b);
	}
	else
		putc(min, q);
	return 0;
}
